File: wasserbilanz.m
%% Übung 1.5 Wasserbilanz und Doppelsummenkurven

%% Daten einlesen und vorbereiten
% Daten liegen im aktuellen Arbeitsverzeichnis von Matlab
% mit eigener Funktion (springt über Header, formatiert Datum)
obst = readWasserbilanzdat('1.5_Wasserbilanz_OBST.csv');
bech = readWasserbilanzdat('1.5_Wasserbilanz_BECH.csv');

% EZG-Flaechen (stehen als km² im Header)
A_obst = 81 * 1e6 ;       % m²
A_bech = 93.4 * 1e6  ;   % m²

% Abfluss von m³/s in mm/h umrechnen, neue Spalte in Tabelle
obst.Q_mm = obst.Q*3600*1000/A_obst ;
bech.Q_mm = bech.Q*3600*1000/A_bech ;


%% Kumulierte Wasserbilanz über ganzen Zeitraum (6 Jahre) zeigen
 subplot(1,2,1);
    plot(obst.Datum, cumsum(obst.N))
    hold on
    plot(obst.Datum, cumsum(obst.Q_mm))
    hold off
    title('Oberstdorf/Stillach')
    xlabel('Datum') 
    ylabel('mm') 
    legend('Niederschlag', 'Abfluss', 'Location','northwest')

  subplot(1,2,2);
    plot(bech.Datum, cumsum(bech.N))
    hold on
    plot(bech.Datum, cumsum(bech.Q_mm))
    hold off
    title('Bechhofen/Wieseth')
    xlabel('Datum') 
    ylabel('mm') 
    legend('Niederschlag', 'Abfluss', 'Location','northwest')
    
%% Doppelsummenkurven für einzelne Jahre bestimmen

    % Wasserwirtschaftsjahre
    obst.wwj = year(obst.Datum);         % Jahr als Zahl
    bech.wwj = year(bech.Datum);         
    % November und Dezember gehören zum nächsten Jahr
    %       -> das Jahr für diese Einträge wird eins weitergezählt
    %       der Index ist für beide Stationen gleich
    novdez_ind = ismember(month(obst.Datum),[11 12]);
    obst.wwj( novdez_ind ) = obst.wwj( novdez_ind )+1;
    bech.wwj( novdez_ind ) = bech.wwj( novdez_ind )+1;
    
    %% Doppelsummenkurve
    % Kumulieren der Niederschläge und Abflüsse
    % normieren mit Jahresniederschlag
    % jeweils pro WWJ - hier wieder für beide gleich   
    
    % Gruppieren der Daten nach WWJ
    [G, wwjs] = findgroups(obst.wwj);

    % Überprüfen, ob beide Reihen gleich lang sind
          if(G ~= findgroups(bech.wwj)) 
              warning("Data not of same length");
          end
      
    % Jahresniederschläge je WWJ
    Ntot_obst = splitapply(@sum, obst.N, G); 
     Ntot_bech = splitapply(@sum, bech.N, G);
    
    % kumulierte Reihen je WWJ
    Nnorm_obst = splitapply(@(x1) {cumsum(x1)}, obst.N, G);
     Qnorm_obst = splitapply(@(x1) {cumsum(x1)}, obst.Q_mm, G);
    Nnorm_bech = splitapply(@(x1) {cumsum(x1)}, bech.N, G);
     Qnorm_bech = splitapply(@(x1) {cumsum(x1)}, bech.Q_mm, G);    

    % normieren: jede Reihe durch jeweiligen Jahresniederschlag teilen
    for jj = 1:length( wwjs)
      Nnorm_obst{jj} = Nnorm_obst{jj}/Ntot_obst(jj);
        Qnorm_obst{jj} = Qnorm_obst{jj}/Ntot_obst(jj);
      Nnorm_bech{jj} = Nnorm_bech{jj}/Ntot_bech(jj);
        Qnorm_bech{jj} = Qnorm_bech{jj}/Ntot_bech(jj);
    end
    
   %% Grafiken
  
  figure       %% macht ggf ein neues Grafikfenster auf
  subplot(1,2,1);
    hold on
    for jj = 1:length( Ntot_obst)
        plot(Nnorm_obst{jj},Qnorm_obst{jj})
    end
    hold off
    
    ylim([0 1]);
    title('Oberstdorf/Stillach')
    legend(string(wwjs),'Location','northwest')
    xlabel('N_{kumuliert} / N_{tot}') 
    ylabel('Q_{kumuliert} / N_{tot}') 
 
  
  subplot(1,2,2);
    hold on
    for jj = 1:length( Ntot_bech)
        plot(Nnorm_bech{jj},Qnorm_bech{jj})
    end
    hold off
    
    ylim([0 1]);
    title('Bechhofen/Wieseth') 
    legend(string(wwjs),'Location','northwest')
    xlabel('N_{kumuliert} / N_{tot}') 
    ylabel('Q_{kumuliert} / N_{tot}') 
 
  % Ende 
    
    

File: wasserbilanz.R
#### Skript Abflussdaten Übung Hydrologie 1
# Jan Wienhöfer 2024

### 1.5 KLimatische Wasserbilanz / Doppelsummenkurven

# Erst Pegel Oberstdorf, unten dann Bechhofen

 ### Daten einlesen
 #  Oberstdorf
  obst <- read.csv("1.5_Wasserbilanz_OBST.csv",
                   skip=7, colClasses=c("character","character","numeric","numeric"),
                   sep=";")
    colnames(obst) <- c("Datum","Zeit","N","Q")

  # Datum als Zeitvariable
  obst$Datum <- as.Date(obst$Datum, "%d.%m.%Y")

  # Einzugsgebietsgröße
  obst.area <- 81e6                 # m²

  # Q in mm transformieren
  obst$Q.mm <- obst$Q*3600*1000/obst.area        #m³/s *3600 s/h *1000 L/m³ / (A_ezg [m²]) =  mm/h

  # kumulierte Summen über Zeit auftragen

  plot(cumsum(N)~Datum, obst,      # Niederschlag
       t="l", col=4,
       ylab="Kumuliert /mm", main="Oberstdorf/Stillach")

   lines(cumsum(Q.mm) ~Datum, obst,  # Abfluss
         col=6)
   legend("topl", c("Niederschlag", "Abfluss"),
          col=c(4,6), lty=1, bty="n")

   ## Doppelsummenkurven

   # aggregieren zu Tagesdaten
   obst <- aggregate(cbind(N, Q.mm) ~ Datum, data = obst, FUN=sum)

   ## in hydrologische Jahre zerlegen  - laut Din ab 1.11.
    # Jahreszahl aus Datum
    obst$hj <- as.numeric( strftime(obst$Datum, "%Y"))
    # Monate Nov und Dez finden und Jahreszahl für diese um eins erhöhen
    obst$hj[strftime(obst$Datum, "%m") %in% c(11,12)] <- obst$hj[strftime(obst$Datum, "%m") %in% c(11,12)]+1

    # Datensatz nach hydrologischen Jahren zerlegen
     hj.obst <- split(obst, obst$hj)

   # Doppelsummenkurven für jedes hyd Jahr bestimmen
   dmc.obst <- lapply(hj.obst, function(x) { cQ <- cumsum(x$Q.mm)    # kumulieren
                                        cN <- cumsum(x$N)       # kumulieren
                                        totN <- sum(x$N)
                                        res <- cbind(cN, cQ)/totN                  # normieren
                                        return(res)
                                        })

   # PLot Doppelsummenkurven
   plot(cQ~cN, dmc.obst[[1]], t="l", ylim=c(0,1),
        main="Oberstdorf Doppelsummenkurve" )
   for(jj in 2:length(dmc.obst)) lines(cQ~cN, dmc.obst[[jj]], col=jj)
     legend("topleft", names(dmc.obst), col=1:length(dmc.obst), lty=1, bty="n")

 #############
 #
 #  Bechhofen

  bech <- read.csv("1.5_Wasserbilanz_bech.csv", skip=7, colClasses=c("character","character","numeric","numeric"), sep=";")
  colnames(bech) <- c("Datum","Zeit","N","Q")
  bech$Datum <- as.Date(bech$Datum, "%d.%m.%Y")

  bech.area <- 93.4e6  # m²

  # Q in mm
  bech$Q.mm <- bech$Q*3600*1000/bech.area   # m³/s *3600 s/h *1000 L/m³ / (A_ezg [m²]) =  mm/h

  plot(cumsum(N)~Datum, bech, t="l", col=4, ylab="Kumuliert/mm", main="Bechhof/Wieseth")
   lines(cumsum(Q.mm) ~Datum, bech, col=6)
   legend("topl", c("Niederschlag", "Abfluss"), col=c(4,6), lty=1, bty="n")

  ## Doppelsummenkurven

  # aggregieren (summieren) zu Tagesdaten
  bech <- aggregate(cbind(N, Q.mm) ~ Datum, data = bech, FUN=sum)

  # in hydrologische Jahre zerlegen  - laut Din ab 1.11.
  bech$hj <- as.numeric(strftime(bech$Datum, "%Y") )
  bech$hj[strftime(bech$Datum, "%m") %in% c(11,12)] <- bech$hj[strftime(bech$Datum, "%m") %in% c(11,12)]+1

  plot(cumsum(N)~Datum, bech, t="l")
   lines(cumsum(Q.mm)~Datum, bech, col=3)

   hj.bech <- split(bech, bech$hj)

   dmc.bech <- lapply(hj.bech, function(x) { cQ <- cumsum(x$Q.mm)    # kumulieren
                                        cN <- cumsum(x$N)       # kumulieren
                                        totN <- sum(x$N)
                                        res <- cbind(cN, cQ)/totN                  # normieren
                                        return(res)
                                        })

  # Plot Double mass curve
   plot(cQ~cN, dmc.bech[[1]], t="l", ylim=c(0, 1) ) # 1.1*max(sapply(dmc.bech, function(x) max(x[, "cQ"]) ))) )
   for(jj in 2:length(dmc.bech)) lines(cQ~cN, dmc.bech[[jj]], col=jj)
   legend("topleft", names(dmc.bech), col=1:length(dmc.bech), lty=1, bty="n")
